airflow/models/taskinstance.py

```diff
@@ -3726,7 +3726,7 @@ def _schedule_downstream_tasks(
             partial_dag = task.dag.partial_subset(
                 task.downstream_task_ids,
                 include_downstream=True,
-                include_upstream=False,
+                include_upstream=True,
```

Comment:
@ephraimbuddy / @uranusjr - I'd be happy for your feedback on this one. It solves the issue, yet it somewhat seems to degenerate the idea of fetching the partial DAG (i.e., due to the dependency in the upstream's trigger rules, you actually fetch the entire branch). Do we need to make this fetching more granular, or on the contrary  simplify it for performance reasonings? (not necessarily as part of this PR)

Reply:
You are right. In fact, I thought about this problem before submitting this PR. I tried to divide the subgraph more accurately. Unfortunately, I have not found a better solution yet.

I looked at the entire scheduling process. When `include_upstream=true`, there may be more unnecessary nodes, but the impact on scheduling is not great, because the list of unrunning nodes will be filtered out in the end, and then the upstream of these nodes will be analyzed to decide whether to schedule.

In addition, the current division method does not obtain the entire dag. For example, when a dag contains multiple non-connected parts, only part of it will be obtained.

I did a simple test, running several different types of dags 10 times under the same conditions (the topology of these dags is relatively complex but the task is very simple, similar to BashOperator's `'echo xxx'`), and then compared the average scheduling time, and finally found that the time consumption has basically not changed.


| dag_id  | task_num | `include_upstream=False` | `include_upstream=True` |
| --- | --- | --- | --- |
| example_branch_operator | 23 | 30.9s | 31.9s |
| example_complex | 37 | 79.4s | 78.8s |

Reply:
Actually, I would love to see if removing partial altogether is not a better solution. There are quite a few problems coming from using partial - manifesting with "schedule_downstream_tasks" failing when DAG could not be serialized - because partial_subset performs serialization internally.

Up till recently such failure in "mini-scheduler" was silently swallowed - it was not printed in the task, but it caused flower's errors and a number of users complained that "tasks succeeded" but the corresponding celery tasks were marked as "failed". I "fixed" it in https://github.com/apache/airflow/pull/41260 by logging an error message - but the message is anyhow not actionable - see https://github.com/apache/airflow/issues/39717 for example.

Your calculation @luoyuliuyin  made me think that maybe this was a premature optimisation and maybe we might simply avoid running partial_subset at all in this case ?

Reply:
cc: @uranusjr @ashb  -> I think you are most familiar with that part of the code.

Reply:
I looked at some of the commit history(https://github.com/apache/airflow/pull/11589, https://github.com/apache/airflow/pull/16289, https://github.com/apache/airflow/pull/18338, https://github.com/apache/airflow/pull/27506, https://github.com/apache/airflow/pull/41260) for this part of the code and found that the parameters of partial_subset have changed several times. It is indeed worth discussing partial_subset.

Reply:
@luoyuliuyin -> out of curiosity could you re-run your benchmarks after removing `partial_subset` ?

Reply:
OK, the test results don't seem to be much different.
![image](https://github.com/user-attachments/assets/829b56ff-31be-4930-b364-ec192648c837)

| dag_id | task_num | include_upstream=False | include_upstream=True | remove partial_subset |
| -- | -- | -- | -- | -- |
| example_branch_operator | 23 | 30.9s | 31.9s | 31.4s |
| example_complex | 37 | 79.4s | 78.8s | 78.8s |